File: src/com/example/tvbrowser/VoiceCommandEngine.java
package com.example.tvbrowser;

public class VoiceCommandEngine {

    public enum ActionType {
        OPEN_URL,
        SEARCH,
        OPEN_BOOKMARKS,
        OPEN_DOWNLOADS,
        OPEN_SETTINGS,
        TOGGLE_CURSOR,
        NEW_TAB,
        CLOSE_TAB,
        RELOAD
    }

    public static class CommandResult {
        public ActionType type;
        public String payload;
        public String spokenText;

        public CommandResult(ActionType type, String payload, String spokenText) {
            this.type = type;
            this.payload = payload;
            this.spokenText = spokenText;
        }
    }

    public static CommandResult parse(String text) {
        if (text == null || text.trim().isEmpty()) {
            return new CommandResult(ActionType.SEARCH, "", "");
        }

        String raw = text.trim();
        String q = raw.toLowerCase();

        // 1. Direct App / Portal Intenti
        if (q.contains("youtube") || q.contains("jutub") || q.contains("ju tub")) {
            return new CommandResult(ActionType.OPEN_URL, "https://www.youtube.com", raw);
        }
        if (q.contains("stream") || q.contains("nexus") || q.contains("film") || q.contains("filmi") || q.contains("serij")) {
            return new CommandResult(ActionType.OPEN_URL, "http://192.168.0.135:3000", raw);
        }
        if (q.contains("tmdb") || q.contains("the movie database") || q.contains("baza filmov")) {
            return new CommandResult(ActionType.OPEN_URL, "https://www.themoviedb.org", raw);
        }
        if (q.equals("google") || q.contains("odpri google")) {
            return new CommandResult(ActionType.OPEN_URL, "https://www.google.com", raw);
        }
        if (q.contains("reddit")) {
            return new CommandResult(ActionType.OPEN_URL, "https://www.reddit.com", raw);
        }
        if (q.contains("github") || q.contains("git hub")) {
            return new CommandResult(ActionType.OPEN_URL, "https://github.com", raw);
        }
        if (q.contains("twitch")) {
            return new CommandResult(ActionType.OPEN_URL, "https://www.twitch.tv", raw);
        }
        if (q.contains("radio") || q.contains("radio garden") || q.contains("glasb")) {
            return new CommandResult(ActionType.OPEN_URL, "https://radio.garden", raw);
        }

        // 2. Tab Management Intenti
        if (q.contains("nov zavihek") || q.contains("novi zavihek") || q.contains("new tab") || q.contains("dodaj zavihek")) {
            return new CommandResult(ActionType.NEW_TAB, null, raw);
        }
        if (q.contains("zapri zavihek") || q.contains("close tab") || q.contains("odstrani zavihek")) {
            return new CommandResult(ActionType.CLOSE_TAB, null, raw);
        }

        // 3. UI Navigation Intenti
        if (q.contains("zaznamk") || q.contains("priljubljen") || q.contains("bookmark")) {
            return new CommandResult(ActionType.OPEN_BOOKMARKS, null, raw);
        }
        if (q.contains("prenos") || q.contains("datotek") || q.contains("download")) {
            return new CommandResult(ActionType.OPEN_DOWNLOADS, null, raw);
        }
        if (q.contains("nastavitv") || q.contains("setting") || q.contains("opcij")) {
            return new CommandResult(ActionType.OPEN_SETTINGS, null, raw);
        }
        if (q.contains("kurzor") || q.contains("miš") || q.contains("pointer") || q.contains("kazalec")) {
            return new CommandResult(ActionType.TOGGLE_CURSOR, null, raw);
        }
        if (q.contains("osveži") || q.contains("reload") || q.contains("ponovno naloži")) {
            return new CommandResult(ActionType.RELOAD, null, raw);
        }

        // 4. Prefix Search / AI Query Stripping
        String searchPayload = raw;
        String[] prefixes = {"poišči", "išči", "najdi", "search", "pokaži", "odpri"};
        for (String p : prefixes) {
            if (q.startsWith(p + " ")) {
                searchPayload = raw.substring(p.length() + 1).trim();
                break;
            }
        }

        return new CommandResult(ActionType.SEARCH, searchPayload, raw);
    }
}

File: src/main/kotlin/com/example/tvbrowser/VoiceCommandEngine.kt
package com.example.tvbrowser

import java.net.URLEncoder

enum class CommandType {
    OPEN_URL,
    SEARCH,
    NEW_TAB,
    CLOSE_TAB,
    OPEN_BOOKMARKS,
    OPEN_DOWNLOADS,
    OPEN_SETTINGS,
    TOGGLE_CURSOR,
    RELOAD
}

data class VoiceResult(val type: CommandType, val payload: String = "")

object VoiceCommandEngine {

    fun parse(spokenText: String?): VoiceResult {
        if (spokenText.isNullOrBlank()) {
            return VoiceResult(CommandType.SEARCH, "")
        }

        val raw = spokenText.trim().lowercase()

        // 1. YouTube & Iskanje Pesmi / Glasbe / Videoposnetkov
        if (raw == "youtube" || raw == "odpri youtube" || raw == "open youtube" || raw == "smarttube" || raw == "odpri smarttube") {
            return VoiceResult(CommandType.OPEN_URL, "https://www.youtube.com")
        }
        if (raw.startsWith("predvajaj ") || raw.startsWith("zavrti ") || raw.startsWith("poslušaj ") ||
            raw.startsWith("pesem ") || raw.startsWith("komad ") || raw.startsWith("spot ") ||
            raw.startsWith("youtube ") || raw.startsWith("poišči na youtube") || raw.startsWith("poišči na youtubu") ||
            raw.contains("na youtube") || raw.contains("na youtubu")) {
            
            var query = raw
            val ytPrefixes = listOf(
                "odpri youtube in poišči ", "odpri youtube in predvajaj ", "odpri youtube ",
                "poišči na youtubu ", "poišči na youtube ", "poišči v youtubu ",
                "predvajaj pesem od ", "predvajaj glasbo od ", "predvajaj spot od ",
                "predvajaj pesem ", "predvajaj komad ", "predvajaj glasbo ", "predvajaj spot ", "predvajaj ",
                "zavrti pesem od ", "zavrti glasbo ", "zavrti spot ", "zavrti ",
                "poslušaj pesem ", "poslušaj glasbo ", "poslušaj ",
                "pesem od ", "pesem ", "komad ", "spot ",
                "youtube ", "smarttube ", "jutub "
            )
            for (p in ytPrefixes) {
                if (query.startsWith(p)) {
                    query = query.substring(p.length).trim()
                    break
                }
            }
            query = query.replace("na youtube", "").replace("na youtubu", "").replace("on youtube", "").trim()
            return if (query.isNotEmpty()) {
                VoiceResult(CommandType.OPEN_URL, "https://www.youtube.com/results?search_query=" + URLEncoder.encode(query, "UTF-8"))
            } else {
                VoiceResult(CommandType.OPEN_URL, "https://www.youtube.com")
            }
        }

        if (raw == "tmdb" || raw == "odpri tmdb") {
            return VoiceResult(CommandType.OPEN_URL, "https://www.themoviedb.org")
        }

        if (raw == "github" || raw == "odpri github") {
            return VoiceResult(CommandType.OPEN_URL, "https://github.com")
        }

        if (raw == "streamnexus" || raw == "odpri streamnexus") {
            return VoiceResult(CommandType.OPEN_URL, "https://www.themoviedb.org/movie")
        }

        // 2. Browser Controls
        if (raw.contains("nov zavihek") || raw.contains("odpri zavihek") || raw.contains("new tab")) {
            return VoiceResult(CommandType.NEW_TAB)
        }
        if (raw.contains("zapri zavihek") || raw.contains("close tab")) {
            return VoiceResult(CommandType.CLOSE_TAB)
        }
        if (raw.contains("zaznamki") || raw.contains("priljubljene") || raw.contains("bookmarks")) {
            return VoiceResult(CommandType.OPEN_BOOKMARKS)
        }
        if (raw.contains("prenosi") || raw.contains("downloads")) {
            return VoiceResult(CommandType.OPEN_DOWNLOADS)
        }
        if (raw.contains("nastavitve") || raw.contains("settings")) {
            return VoiceResult(CommandType.OPEN_SETTINGS)
        }
        if (raw.contains("kazalec") || raw.contains("miška") || raw.contains("kurzor") || raw.contains("cursor")) {
            return VoiceResult(CommandType.TOGGLE_CURSOR)
        }
        if (raw.contains("osveži") || raw.contains("ponovno naloži") || raw.contains("reload")) {
            return VoiceResult(CommandType.RELOAD)
        }

        // 3. Glavni privzeti iskalnik: VEDNO GOOGLE za vsa iskanja ("search for rtv", "poišči...", itd.)
        var cleanQuery = raw
        val prefixes = listOf("search for ", "search ", "poišči ", "išči ", "najdi ", "google ")
        for (p in prefixes) {
            if (cleanQuery.startsWith(p, ignoreCase = true)) {
                cleanQuery = cleanQuery.substring(p.length).trim()
                break
            }
        }
        return VoiceResult(CommandType.SEARCH, cleanQuery.ifEmpty { raw })
    }
}
